### PR TITLE
Completion feature

### PR DESCRIPTION
**Pull request by YevhenKap**
I have add simple completion feature. It will work for all variables, functions, classes in scope and for standard types (like Array, Map, Math and so on). Unfortunately properties and methods of objects is not completes yet :(

### CODE REVIEW

**Comment by YevhenKap** [comment]
Maybe you can add some advise?)

**Comment by JSMonk** [comment]
Oh, cool. Thank you a lot. I'm currently sitting with my family, but I will check and write some feedback about it tomorrow. Thank you a lot for your contribution ^_^.

**Comment by YevhenKap** [comment]
It is an honor to be involved in the development of a great product. 😊 I hope that I do not make it bad because of my lack of knowledge.
You are cool man) And you start great project)

File: packages/language-server/server/completion/completion.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  getPositionedModuleScopeTypes,
+} = require("../validation/code_validation");
+
+async function onCompletion(completionParams) {
+  const types = getPositionedModuleScopeTypes();
+
+  const completionItems = [];
+
+  if (types && types.body) {
+    completionItems.push(...buildCompletionItem(types, 1));
```
Comment (JSMonk) [inline packages/language-server/server/completion/completion.js:12]: Could we there make something like:
```javascript
return types !== undefined && types.body !== undefined 
  ? types.bodybuildCompletionItem(types, 1);
  : []; 
```
And at the like 15

File: packages/language-server/server/completion/completion.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  for (const [name, variableInfo] of scope.body.entries()) {
+    completionItems.push({
+      label: name,
+      kind: getCompletionKind(variableInfo),
+      data: dataIndex++,
+    });
+  }
+
+  if (scope.parent !== null) {
+    completionItems.push(...buildCompletionItem(scope.parent, dataIndex++));
```
Comment (JSMonk) [inline packages/language-server/server/completion/completion.js:30]: Same thing there:
```javascript
return scope.parent !== null 
  ? [...completionItems,  ...buildCompletionItem(scope.parent, dataIndex + 1)]
  : completionItems

File: packages/language-server/server/completion/completion_resolve.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
+
+function onCompletionResolve(item) {
+  item.detail = 'Type';
+  item.documentation = 'JavaScript documentation';
+  return item;
```
Comment (JSMonk) [inline packages/language-server/server/completion/completion_resolve.js:6]: And there:
```
return {
  ...item,
  detail: "Type",
 documentation: "JavaScript Documentatnion"
};
```
I really like immutable style of programming, so, try to keep the style in our projects 😄

File: packages/language-server/server/utils/document_ast.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+const babylon = require("@babel/parser");
```
Comment (JSMonk) [inline packages/language-server/server/utils/document_ast.js:1]: And could I ask you to give a name to documents in kebab case: `document-ast.js`?

File: packages/language-server/server/validation/typings.js [hunk 5]
```diff
@@ -40,14 +34,8 @@ async function getModuleAST(currentModulePath) {
...
-    async (modulePath, isTypings) => {
-      let moduleContent = await readFile(modulePath, "utf8");
-      moduleContent =
-        extname(modulePath) === ".json"
-          ? wrapJSON(moduleContent)
-          : moduleContent;
-      const config = isTypings ? dtsrc : babelrc;
-      return babylon.parse(moduleContent, config);
+    (modulePath, isTypings) => {
+      return getBabylonAST(modulePath, null, isTypings);
```
Comment (JSMonk) [inline packages/language-server/server/validation/typings.js:38]: And there, we can drop `return` and `{}`

File: packages/language-server/server/completion/completion.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  return completionItems;
+}
+
+/**
+ * Find type kind of variable.
+ */
+function getCompletionKind(variableInfo) {
+  const kinds = [
```
Comment (JSMonk) [inline packages/language-server/server/completion/completion.js:40]: Seems like we can move `kinds` at the top-level scope.

File: packages/language-server/server/completion/completion_resolve.js (hunk 3 above)
Comment (YevhenKap) [reply packages/language-server/server/completion/completion_resolve.js:6]: Okay, I will try)

**Comment by YevhenKap** [comment]
I fix my mistakes and add completion for inner methods and properties of objects. Also methods and properties of global objects (like Array, Map) completes too now. But there is speed of server problem :( Extension needs an optimization.

**Comment by JSMonk** [comment]
@YevhenKap . Could you please describe the feature? I don't work with compilation in VS Code before, so, I want to understand how it works and what will we get 😄

**Comment by YevhenKap** [comment]
Of course))) But not feature is about autocompletion, not compilation) When users start type word, VSCode try find variable which starts with typed character sequence and suggest it, also it will suggests object methods and properties if there are any. [Here is the video](https://drive.google.com/open?id=1HtAgkGLLQW7FtD1Mn3ibLlj4NcWt8E-m).

**Comment by YevhenKap** [comment]
I add cache for global types. So type variables for completion must be found a little faster. Also I rewrite small chunks of code, so it looks now cleaner. I hope :)

**Comment by JSMonk** [comment]
Thank you a lot for your work. Great job.

**Pull request merged**